cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/MetricStructuredTable.java

```diff
@@ -192,6 +192,48 @@ public CloseableIterator<StructuredRow> scan(Field<?> index) throws InvalidField
     }
   }
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> filterIndex)
+    throws InvalidFieldException, IOException {
+    try {
+      CloseableIterator<StructuredRow> result;
+      if (!emitTimeMetrics) {
```

Comment:
there is a lot of repeated code in these scan methods, it would be good to refactor the repeat code into a common method. Something like:

```

public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> filterIndex) {
  return scan(() -> structuredTable.scan(keyRange, limit, filterIndex), "index.range.scan.");
}

private interface ScanFunction {
  CloseableIterator<StructuredRow> scan() throws InvalidFieldException, IOException;
}

private CloseableIterator<StructuredRow> scan(ScanFunction scanFunc, String metricNamePrefix) {
  try {
    CloseableIterator<StructuredRow> result;
    if (!emitTimeMetrics) {
      result = scanFunc.scan();
    } else {
      long curTime = System.nanoTime();
      result = scanFunc.scan();
      long duration = ...
      ...;
    }
    metricsCollector.increment(...);
    return result;
  } catch (Exception e) {
    metricsCollector.increment(metricPrefix + metricNamePrefix + "error", 1L);
  }
}
```

Reply:
nice suggestion! Working on it now